File: implementations/csharp/Agent.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Azure.AI.OpenAI;

namespace AgentPlatform
{
    public class Agent
    {
        private readonly OpenAIClient _client;
        private readonly string _model;
        private readonly List<ChatRequestMessage> _messages = new();
        private readonly ToolRegistry _registry;

        public Agent(string apiKey, ToolRegistry registry, string model = "gpt-4-turbo")
        {
            _client = new OpenAIClient(apiKey, new OpenAIClientOptions());
            _model = model;
            _registry = registry;
            InitializeSystemPrompt();
        }

        private void InitializeSystemPrompt()
        {
            var systemPrompt = "You are a helpful AI assistant. You have access to various tools. " +
                               "When asked to perform a task, use the tools to gather information and take actions before answering.";
            _messages.Add(new ChatRequestSystemMessage(systemPrompt));
        }

        public async Task<string> RunAsync(string userInput)
        {
            Console.WriteLine($"\nUser: {userInput}");
            _messages.Add(new ChatRequestUserMessage(userInput));

            while (true)
            {
                var options = new ChatCompletionsOptions(_model)
                {
                    Temperature = 0.7f
                };

                foreach (var msg in _messages)
                {
                    options.Messages.Add(msg);
                }

                var tools = _registry.GetToolSchemas();
                foreach (var tool in tools)
                {
                    options.Tools.Add(new ChatCompletionsFunctionToolDefinition(tool));
                }

                var response = await _client.GetChatCompletionsAsync(options);
                var choice = response.Value.Choices[0];
                var message = choice.Message;

                if (message.ToolCalls != null && message.ToolCalls.Count > 0)
                {
                    var assistantMessage = new ChatRequestAssistantMessage(message.Content);
                    
                    foreach (var toolCall in message.ToolCalls)
                    {
                        if (toolCall is ChatCompletionsFunctionToolCall functionCall)
                        {
                            assistantMessage.ToolCalls.Add(functionCall);
                            var name = functionCall.Name;
                            var args = functionCall.Arguments;
                            
                            Console.WriteLine($"  [Tool Execution] {name}({args})");
                            
                            var result = await _registry.ExecuteToolAsync(name, args);
                            Console.WriteLine($"  [Tool Result] {result}");
                            
                            _messages.Add(new ChatRequestToolMessage(result, functionCall.Id));
                        }
                    }
                    // Insert the assistant message before the tool results
                    _messages.Insert(_messages.Count - message.ToolCalls.Count, assistantMessage);
                    
                    continue; // Loop again to send tool results to LLM
                }

                var finalText = message.Content ?? string.Empty;
                Console.WriteLine($"Assistant: {finalText}");
                _messages.Add(new ChatRequestAssistantMessage(finalText));
                return finalText;
            }
        }
    }
}

File: implementations/fsharp/src/Skight.AgentPlatform.FSharp/Agent.fs
namespace Skight.AgentPlatform.FSharp

open System
open System.Collections.Generic
open System.Text.Json
open System.Threading
open Azure.AI.OpenAI

/// Agent class wrapper around the pure functional AgentPipeline
type Agent(apiKey: string, registry: ToolRegistry, config: AgentConfig, ?endpoint: string, ?jwtToken: string) =
    let options = OpenAIClientOptions()
    do
        match jwtToken with
        | Some t when not (String.IsNullOrEmpty t) ->
            options.AddPolicy(BearerTokenPolicy(t), Azure.Core.HttpPipelinePosition.PerCall)
        | _ -> ()

    let client =
        match endpoint with
        | Some ep when not (String.IsNullOrEmpty ep) ->
            OpenAIClient(Uri(ep), Azure.AzureKeyCredential(apiKey), options)
        | _ ->
            OpenAIClient(apiKey, options)

    let (|FunctionToolCall|_|) (tc: ChatCompletionsToolCall) =
        match tc with
        | :? ChatCompletionsFunctionToolCall as fnCall -> Some fnCall
        | _ -> None

    let toDomainResponse (responseMessage: ChatResponseMessage) : LlmTurnResponse =
        let content = if isNull responseMessage.Content then "" else responseMessage.Content
        let toolCalls =
            if isNull responseMessage.ToolCalls then
                []
            else
                responseMessage.ToolCalls
                |> Seq.choose (function
                    | FunctionToolCall fnCall ->
                        match ToolCallId.create fnCall.Id, ToolName.create fnCall.Name with
                        | Ok id, Ok name -> 
                            Some {
                                Id = id
                                Name = name
                                ArgumentsJson = fnCall.Arguments
                            }
                        | _ -> None
                    | _ -> None)
                |> Seq.toList

        {
            Content = content
            ToolCalls = toolCalls
        }

    // Standard default Azure.AI.OpenAI LLM caller implementation
    let defaultLlmCaller : LlmCaller =
        fun schemas msgs ->
            async {
                let requestMessages = msgs |> List.map SdkAdapter.toChatRequestMessage
                let reqOptions = ChatCompletionsOptions(config.Model, requestMessages)
                reqOptions.Temperature <- Nullable(0.7f)
                for schema in schemas do
                    reqOptions.Tools.Add(ChatCompletionsFunctionToolDefinition(SdkAdapter.toFunctionDefinition schema))
                try
                    let! resp = client.GetChatCompletionsAsync(reqOptions) |> Async.AwaitTask
                    let completions = resp.Value
                    if completions.Choices.Count = 0 then
                        return Error NoChoicesReturned
                    else
                        return Ok (toDomainResponse completions.Choices.[0].Message)
                with ex ->
                    return Error (ApiCallFailed ex.Message)
            }

    let defaultStreamingCallerFactory (cancellationToken: CancellationToken) (onChunk: StreamChunk -> unit) : StreamingLlmCaller =
        fun schemas msgs ->
            async {
                try
                    let stream = SdkAdapter.streamLlmResponseWithCallback client config schemas msgs cancellationToken onChunk
                    return Ok stream
                with ex ->
                    return Error (ApiCallFailed ex.Message)
            }

    let systemPrompt =
        "You are a helpful AI assistant. You have access to tools including file system operations and terminal command execution. " +
        "When asked to perform a task, inspect tool descriptions and use appropriate tools to gather information and take actions before answering."

    member _.SystemPrompt : string = systemPrompt

    member _.DefaultLlmCaller : LlmCaller = defaultLlmCaller

    member _.CreateInitialSession() : AgentSessionState =
        AgentSession.initialize systemPrompt

    member _.RunPureAsync(
        userInput: string,
        sessionState: AgentSessionState,
        ?customLlmCaller: LlmCaller,
        ?customExecutor: ToolExecutor,
        ?registeredSchemas: ToolSchema list,
        ?registeredNamesSet: Set<ToolName>
    ) : Async<TurnResult * AgentSessionState> =
        async {
            let activeLlmCaller = defaultArg customLlmCaller defaultLlmCaller
            let activeExecutor = defaultArg customExecutor registry.AsExecutor
            let schemas = defaultArg registeredSchemas (registry.GetToolSchemas())
            let namesSet = defaultArg registeredNamesSet (registry.GetRegisteredNames() |> Set.ofList)

            return! AgentRunner.runTurnAsync activeLlmCaller activeExecutor config userInput sessionState schemas namesSet
        }

    member _.RunPureStreamingAsync(
        userInput: string,
        sessionState: AgentSessionState,
        onChunk: StreamChunk -> unit,
        ?customStreamingLlmCallerFactory: (CancellationToken -> (StreamChunk -> unit) -> StreamingLlmCaller),
        ?customExecutor: ToolExecutor,
        ?registeredSchemas: ToolSchema list,
        ?registeredNamesSet: Set<ToolName>,
        ?cancellationToken: CancellationToken
    ) : Async<TurnResult * AgentSessionState> =
        async {
            let activeFactory = defaultArg customStreamingLlmCallerFactory defaultStreamingCallerFactory
            let activeExecutor = defaultArg customExecutor registry.AsExecutor
            let schemas = defaultArg registeredSchemas (registry.GetToolSchemas())
            let namesSet = defaultArg registeredNamesSet (registry.GetRegisteredNames() |> Set.ofList)
            let ct = defaultArg cancellationToken CancellationToken.None

            return! AgentRunner.runTurnStreamingAsync activeFactory activeExecutor config userInput sessionState schemas namesSet onChunk ct
        }
